### PR TITLE
Fix cached decorations

### PR DESCRIPTION
**Pull request by pmconne**
`Viewport.computeViewRange` - called from clip decorator - was calling `setupFromView` (courtesy of Nate Rex), which invalidated scene and therefore all cached decorations. That is bad while we are in the process of populating that cache.

A decorator's `decorate` method should decorate, not modify the viewport. Removed `setupFromView` call.

Added DecorationsCache and set a flag on it for the duration of `ScreenViewport.addDecorations` to prohibit removal.

Moved most of the cached-decorations-related APIs to DecorationsCache; the cache itself moved from `Viewport` to `ScreenViewport`.

[Work item.](https://bentleycs.visualstudio.com/iModelTechnologies/_workitems/edit/549987/)

### CODE REVIEW

**Comment by markschlosseratbentley** [comment]
@pmconne I gave it a read-through and it looks good - (your comment about tests - I agree, sounds good to add).  Tomorrow I'll try to test this branch out.

**Review by bbastings — APPROVED** [review]
Looks good.

File: full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (markschlosseratbentley) [inline full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts:1]: Looks like there was some fairly basic testing in Decorations.test.ts for cached decorations - it's great to have this very thorough test now.

Comment (pmconne) [reply full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts:1]: Most of it was already there, I just renamed the file and added test for my changes.

File: common/api/imodeljs-frontend.api.md [hunk 2]
```diff
@@ -8307,7 +8325,7 @@ export class ScreenViewport extends Viewport {
     openToolTip(message: HTMLElement | string, location?: XAndY, options?: ToolTipOptions): void;
     readonly parentDiv: HTMLDivElement;
     // @internal (undocumented)
-    pickCanvasDecoration(pt: XAndY): CanvasDecoration | undefined;
```
Comment (bbastings) [inline common/api/imodeljs-frontend.api.md:8310]: Is this import here weird?

Comment (pmconne) [reply common/api/imodeljs-frontend.api.md:8310]: extract-api is weird. I don't worry about it unless the PR build complains.

File: full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts (hunk 1 above)
Comment (markschlosseratbentley) [reply full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts:1]: @pmconne Separate files makes it clearer... nice.

Comment (markschlosseratbentley) [reply full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts:1]: @pmconne Oh, looks like we don't have a Decorations.test.ts or anything which just has decorations tested.  I'm sure we do somewhere have more through decoration testing, right?

Comment (pmconne) [reply full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts:1]: See prev reply...I renamed `Decorations.test.ts` to `CachedDecorations.test.ts`, did some minor cleanup of existing tests, and added my new tests.

Comment (markschlosseratbentley) [reply full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts:1]: I gotcha now - I initially looked at this file wrong.  Thanks for clearing me up.
Not for this PR, but maybe we could think about testing decorators in some kind of more automated way (rendering one, checking the pixels) like we do for mirukuru, etc.  Might catch some more stuff.

**Review by markschlosseratbentley — APPROVED** [review]
All looks good - I built & tested on my local machine and decorators all seem to be working properly.

File: full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts (hunk 1 above)
Comment (pmconne) [reply full-stack-tests/core/src/frontend/standalone/CachedDecorations.test.ts:1]: More tests == more better.
Also for separate PR: tests like these ones that use a Viewport but don't care about a specific iModel should be moved to imodeljs-frontend and use a BlankConnection. The annoying part of that is we'd want to be able to use TestViewport.ts which is currently in core-full-stack-tests.

**Review by swwilson-bsi — APPROVED** [review]

**Review by kabentley — APPROVED** [review]

**Comment by pmconne** [comment]
@Mergifyio backport release/2.12.x

**Pull request merged**